core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2065,7 +2063,79 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
   ]: Concurrent, O2](
       maxConcurrent: Int
   )(f: O => F2[O2]): Stream[F2, O2] =
-    map(o => Stream.eval(f(o))).parJoin(maxConcurrent)
+    if (maxConcurrent == 1) evalMap(f)
+    else {
+      assert(maxConcurrent > 0, "maxConcurrent must be > 0, was: " + maxConcurrent)
+
+      // One is taken by inner stream read.
+      val concurrency = if (maxConcurrent == Int.MaxValue) Int.MaxValue else maxConcurrent + 1
+      val action =
+        (
+          Semaphore[F2](concurrency.toLong),
+          Channel.bounded[F2, O2](concurrency),
+          Ref[F2].of(none[Either[NonEmptyList[Throwable], Unit]]),
+          Deferred[F2, Either[Throwable, Unit]]
+        ).mapN { (semaphore, channel, result, stopReading) =>
+          val releaseAndCheckCompletion =
+            semaphore.release *>
+              semaphore.available
+                .product(result.get)
+                .flatMap { case (available, completion) =>
+                  channel.close.whenA(completion.nonEmpty && available == concurrency)
+                }
+
+          val succeed =
+            result.update {
+              case None  => ().asRight.some
+              case other => other
+            }
+
+          val cancelled = stopReading.complete(().asRight) *> succeed
+
+          def failed(ex: Throwable) =
+            stopReading.complete(().asRight) *>
+              result.update {
+                case Some(Left(nel)) => nel.prepend(ex).asLeft.some
+                case _               => NonEmptyList.one(ex).asLeft.some
+              }
+
+          val completeStream =
+            Stream.force {
+              result.get.map {
+                case Some(Left(nel)) => Stream.raiseError[F2](CompositeFailure.fromNel(nel))
+                case _               => Stream.empty
+              }
+            }
+
+          val pullExecAndOutput =
+            Stream.exec(semaphore.acquire) ++
+              interruptWhen(stopReading)
+                .evalMap { el =>
```

Comment:
Are we interested on "cutting" the streams of units this `evalMap` would give? Otherwise, perhaps we can just do a `foreach`, and get a `Stream[F2, INothing]` from it.